libs/contrib/Data/List/Equalities.idr

```diff
@@ -0,0 +1,60 @@
+module Data.List.Equalities
+
+import Data.List
+
+%default total
+
+||| (::) is injective
+export
+consInjective : {x : a} -> {xs : List a} -> {y : b} -> {ys : List b} ->
+                (x :: xs) = (y :: ys) -> (x = y, xs = ys)
+consInjective Refl = (Refl, Refl)
+
+||| Two lists are equal, if their heads are equal and their tails are equal.
+consCong2 : {x : a} -> {xs : List a} -> {y : b} -> {ys : List b} ->
+            x = y -> xs = ys -> x :: xs = y :: ys
+consCong2 Refl Refl = Refl
+
+||| Equal non-empty lists should result in equal componens after destructuring 'snoc'.
+export
+snocCong2 : {x : a} -> {xs : List a} -> {y : b} -> {ys : List b} ->
+            (xs ++ [x]) = (ys ++ [y]) -> (xs = ys, x = y)
+
+snocCong2 {xs = []} {ys = []} Refl = (Refl, Refl)
+snocCong2 {xs = []} {ys = z :: ys} prf = snocCong2 prf
+snocCong2 {xs = w :: xs} {ys = []} prf = snocCong2 prf
+snocCong2 {xs = w :: xs} {ys = z :: ys} prf =
+  let
+    (wEqualsZ, xsSnocXEqualsYsSnocY) = consInjective prf
+    (xsEqualsYS, xEqualsY) = snocCong2 xsSnocXEqualsYsSnocY
+   in (consCong2 wEqualsZ xsEqualsYS, xEqualsY)
+
+ ||| Appending pairwise equal lists gives equal lists
+appendCong2 : {x1 : List a} -> {x2 : List a} ->
+              {y1 : List b} -> {y2 : List b} ->
+              x1 = y1 -> x2 = y2 -> x1 ++ x2 = y1 ++ y2
+appendCong2 {x1=[]} {y1=(_ :: _)} Refl _ impossible
+appendCong2 {x1=(_ :: _)} {y1=[]} Refl _ impossible
+appendCong2 {x1=[]} {y1=[]} _ eq2 = eq2
+appendCong2 {x1=(_ :: _)} {y1=(_ :: _)} eq1 eq2 =
+  consCong2
+    (fst $ consInjective eq1)
+    (appendCong2 (snd $ consInjective eq1) eq2)
```

Comment:
Instead of calling `fst` and then `snd` on `consInjective eq1`,
why not use `let (hd, tl) = consInjective eq1 in`?

Reply:
Updated PR. Thanks. This was code borrowed from Idris1.